File: C#/P1271/Solution.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.IO;

public class Solution
{
    private void RadixSort(int[] nums, int max)
    {
        if (nums == null || nums.Length == 0)
        {
            return;
        }

        int radix = 1;
        int d = 10;
        int[] aux = new int[nums.Length];

        while (max / radix > 0)
        {
            int[] counts = new int[d];
            for (int i = 0; i < nums.Length; i++)
            {
                counts[nums[i] / radix % d]++;
            }

            for (int i = 1; i < d; i++)
            {
                counts[i] += counts[i - 1];
            }

            for (int i = nums.Length - 1; i >= 0; i--)
            {
                aux[--counts[nums[i] / radix % d]] = nums[i];
            }

            for (int i = 0; i < nums.Length; i++)
            {
                nums[i] = aux[i];
            }

            radix *= d;
        }
    }

    // TLE
    public static void Main(string[] args)
    {
        string line = Console.ReadLine();
        string[] parts = line.Split(' ');
        int n = Convert.ToInt32(parts[0]);
        int m = Convert.ToInt32(parts[1]);
        int[] nums = new int[m];
        parts = Console.ReadLine().Split(' ');
        for (int i = 0; i < m; i++)
        {
            nums[i] = Convert.ToInt32(parts[i]);
        }

        var solution = new Solution();
        solution.RadixSort(nums, n);
        if (m > 0)
        {
            for (int i = 0; i < m - 1; i++)
            {
                System.Console.Write("{0} ", nums[i]);
            }

            System.Console.WriteLine($"{nums[m - 1]}");
        }
    }
}

File: C/P1271/Solution.c
#include <stdio.h>

#define SIZE (2000000 + 10)
#define MAX(a, b) ((a) > (b) ? (a) : (b))

int aux[SIZE];
int nums[SIZE];

void radix_sort(int* nums, int m)
{
    int radix = 1;
    int d = 10;
    int max = 0;
    for (int i = 0; i < m; i++) 
    {
        max = MAX(max, nums[i]);
    }

    while (max / radix > 0)
    {
        int count[10] = {0};
        for (int i = 0; i < m; i++)
        {
            count[(nums[i] / radix) % d]++;
        }

        for (int i = 1; i < d; i++)
        {
            count[i] += count[i - 1];
        }

        for (int i = m - 1; i >= 0; i--)
        {
            int idx = count[(nums[i] / radix) % 10];
            aux[--count[(nums[i] / radix) % 10]] = nums[i];
        }

        for (int i = 0; i < m; i++)
        {
            nums[i] = aux[i];
        }

        radix *= d;
    }
}

int main()
{
    int n, m;
    scanf("%d %d", &n, &m);
    for (int i = 0; i < m; i++)
    {
        scanf("%d", &nums[i]);
    }

    radix_sort(nums, m);
    if (m > 0)
    {
        for (int i = 0; i < m - 1; i++)
        {
            printf("%d ", nums[i]);
        }

        printf("%d\n", nums[m - 1]);
    }

    return 0;
}